RPC Meta device support

pbelevich opened the pull request:
We want to support meta device in PyTorch RPC framework. One of possible usecases is creating a large model on meta device for further fx tracing:

1. create a large model on meta device on master node which doesn't have enough memory to materialize the model.
2. trace and split it to submodules using PiPPy
3. send each submodule(on meta device) over RPC to the corresponding workers
4. move those submodules to cuda device(materialize/initialize probably using torchdistx)

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #76882

pbelevich committed:
RPC Meta device support

[ghstack-poisoned]

pbelevich committed:
Update on "RPC Meta device support"




[ghstack-poisoned]

pbelevich committed:
Update on "RPC Meta device support"




[ghstack-poisoned]

pbelevich committed:
Update on "RPC Meta device support"




[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -1423,14 +1423,16 @@ std::vector<c10::Device> TensorPipeAgent::getDevicesForRemote(
           t.device());
     }
-    return {};
+    return getDevicesForTensors(message.tensors(), {{c10::kCPU, c10::kCPU}, {c10::kMeta, c10::kMeta}}, errStr);
```
do we need to add checks somewhere that meta -> not-meta or not-meta -> meta is an error?

mrshenli replied:
do we have to specify cpu and meta here? Or will it work if we default to {CPU->CPU, Meta->Meta} if device map is not specified?

pbelevich replied:
reverted

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -1423,14 +1423,16 @@ std::vector<c10::Device> TensorPipeAgent::getDevicesForRemote(
   } else {
-    return getDevicesForTensors(message.tensors(), iter->second, errStr);
+    DeviceMap mergedMap = iter->second;
+    mergedMap.insert({{c10::kCPU, c10::kCPU}, {c10::kMeta, c10::kMeta}});
```
does it make sense to insert meta -> meta by default into the device map instead of patching it here?

pbelevich replied:
reverted

mrshenli commented on c10/core/Backend.h:
```diff
@@ -107,6 +108,8 @@ static inline Backend dispatchKeyToBackend(DispatchKey t) {
     return Backend::QuantizedXPU;
   } else if (t == DispatchKey::HPU || t == DispatchKey::AutogradHPU) {
     return Backend::HPU;
+  } else if (t == DispatchKey::Meta) {
```
cc @ezyang @jbschlosser

pbelevich replied:
reverted

mrshenli reviewed (commented):
Thanks for adding the support. Is the TensorPipe update intentional? If not, share we do submodule update in a dedicated PR? Thanks!

ezyang commented on c10/core/Backend.h:
```diff
@@ -53,7 +53,8 @@ enum class Backend {
   Lazy,
-  NumOptions
+  NumOptions,
+  Meta
```
this looks wrong, Meta should be before NumOptions

ezyang commented:
What's the intended use case for sending meta tensors over RPC? I have a hard time imagining what its useful for.

pbelevich commented:
> What's the intended use case for sending meta tensors over RPC? I have a hard time imagining what its useful for.

@ezyang 
1. create a large(really large, I mean REALLY LARGE) model on meta device on master node which doesn't have enough memory to materialize the model.
2. trace and split it to submodules using PiPPy
3. send each submodule(on meta device) over RPC to the corresponding workers
4. move those submodules to cuda device(materialize/initialize probably using torchdistx)
5. profit!

cc @jamesr66a @kwen2501

ezyang commented:
Is submodule update intentional?

ezyang commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -113,7 +113,7 @@ std::vector<c10::Device> getDevicesOfTensors(
   std::vector<bool> indexBitset;
   for (const torch::Tensor& tensor : tensors) {
-    if (!tensor.is_cpu()) {
+    if (!tensor.is_cpu() and !tensor.is_meta()) {
```
and is not portable C++, use &&

pbelevich replied:
addressed

ezyang commented on torch/csrc/distributed/rpc/tensorpipe_utils.cpp:
```diff
@@ -91,11 +91,54 @@ class TensorpipeCpuConverter : public TensorpipeDeviceTypeConverter {
 C10_REGISTER_TENSORPIPE_DEVICE_TYPE_CONVERTER(CPU, TensorpipeCpuConverter);
 
+struct MetaBuffer {
+  void* ptr{nullptr};
```
This field is never written to, it seems useless?

pbelevich replied:
agree, it is reverted

ezyang commented on torch/csrc/distributed/rpc/tensorpipe_utils.cpp:
```diff
@@ -91,11 +91,54 @@ class TensorpipeCpuConverter : public TensorpipeDeviceTypeConverter {
+      tensorpipe::Message& message) const override {
+    MetaBuffer buffer;
+    tensorpipe::Message::Tensor tensor;
+    tensor.buffer = buffer;
```
If the buffer length is zero, do you need to assign it at all? (I'm also not terribly familiar with tensorpipe::Message::Tensor, so I'm not sure how you could be assigning a buffer that TensorPipe ostensibly knows nothing about to it)

pbelevich replied:
agree, it is reverted

pbelevich committed:
Update on "RPC Meta device support"

We want to support meta device in PyTorch RPC framework. One of possible usecases is creating a large model on meta device for further fx tracing:

1. create a large model on meta device on master node which doesn't have enough memory to materialize the model.
2. trace and split it to submodules using PiPPy
3. send each submodule(on meta device) over RPC to the corresponding workers
4. move those submodules to cuda device(materialize/initialize probably using torchdistx)




[ghstack-poisoned]

pbelevich committed:
Update on "RPC Meta device support"

We want to support meta device in PyTorch RPC framework. One of possible usecases is creating a large model on meta device for further fx tracing:

1. create a large model on meta device on master node which doesn't have enough memory to materialize the model.
2. trace and split it to submodules using PiPPy
3. send each submodule(on meta device) over RPC to the corresponding workers
4. move those submodules to cuda device(materialize/initialize probably using torchdistx)




[ghstack-poisoned]

pbelevich committed:
Update on "RPC Meta device support"

We want to support meta device in PyTorch RPC framework. One of possible usecases is creating a large model on meta device for further fx tracing:

1. create a large model on meta device on master node which doesn't have enough memory to materialize the model.
2. trace and split it to submodules using PiPPy
3. send each submodule(on meta device) over RPC to the corresponding workers
4. move those submodules to cuda device(materialize/initialize probably using torchdistx)




[ghstack-poisoned]

pbelevich committed:
Update on "RPC Meta device support"

We want to support meta device in PyTorch RPC framework. One of possible usecases is creating a large model on meta device for further fx tracing:

1. create a large model on meta device on master node which doesn't have enough memory to materialize the model.
2. trace and split it to submodules using PiPPy
3. send each submodule(on meta device) over RPC to the corresponding workers
4. move those submodules to cuda device(materialize/initialize probably using torchdistx)




[ghstack-poisoned]

pbelevich committed:
Update on "RPC Meta device support"

We want to support meta device in PyTorch RPC framework. One of possible usecases is creating a large model on meta device for further fx tracing:

1. create a large model on meta device on master node which doesn't have enough memory to materialize the model.
2. trace and split it to submodules using PiPPy
3. send each submodule(on meta device) over RPC to the corresponding workers
4. move those submodules to cuda device(materialize/initialize probably using torchdistx)




[ghstack-poisoned]

pbelevich committed:
Update on "RPC Meta device support"

We want to support meta device in PyTorch RPC framework. One of possible usecases is creating a large model on meta device for further fx tracing:

1. create a large model on meta device on master node which doesn't have enough memory to materialize the model.
2. trace and split it to submodules using PiPPy
3. send each submodule(on meta device) over RPC to the corresponding workers
4. move those submodules to cuda device(materialize/initialize probably using torchdistx)




[ghstack-poisoned]

ezyang commented on torch/csrc/distributed/rpc/tensorpipe_utils.cpp:
```diff
@@ -304,7 +311,7 @@ c10::intrusive_ptr<Message> tensorpipeDeserialize(
       tensorReadFunc,
       {},
-      /* use_storage_device*/ true);
+      /* use_storage_device*/ false);
```
What's going on here?

pbelevich replied:
reverted, thanks!

ezyang commented on torch/csrc/distributed/rpc/tensorpipe_utils.cpp:
```diff
@@ -177,9 +181,12 @@ std::tuple<tensorpipe::Message, TensorpipeWriteBuffers> tensorpipeSerialize(
-  tpMessage.tensors.reserve(tensorDataVec.size());
-  for (const auto i : c10::irange(tensorDataVec.size())) {
-    const torch::Tensor& tensor = tensorDataVec[i];
+  std::vector<torch::Tensor> nonMetaTensorDataVec;
```
This seems very wasteful. Why not just manually take care of the `i` increments in the loop below so you don't have to allocate a temporary vector (with all refcount bumps). I would care less if this only affected meta tensor code but this pessimizes all cases.

pbelevich replied:
addressed

ezyang commented on torch/csrc/distributed/rpc/tensorpipe_utils.cpp:
```diff
@@ -163,12 +163,16 @@ std::tuple<tensorpipe::Message, TensorpipeWriteBuffers> tensorpipeSerialize(
         static_cast<const char*>(buf) + sz);
     return sz;
+  }, nullptr, nullptr, nullptr, [&](const at::Tensor& t) -> std::string {
+    return std::to_string(!t.is_meta() ? non_meta_idx++ : meta_idx++);
```
this seems a bit finely balanced but I'm not sure about a good alternative. It seems like it would be better if the pickler natively understood meta tensors, so that you don't have to stub them out here.

pbelevich replied:
I was thinking about this, but it will require pickler/unpickler to track how many meta/non-meta tensors serialized so far and I decided that it is beyond the scope of this PR

mrshenli replied:
+1 This indeed looks a bit hacky. Shall we create an issue to track this (i.e., adding Meta tensor support to jit pickler/unpickler)? I assume we can treat it as if it is a zero-sized tensor?

ezyang commented:
This seems basically OK, deferring final review to whoever is more familiar with the distributed codebase here.

pbelevich committed:
Update on "RPC Meta device support"

We want to support meta device in PyTorch RPC framework. One of possible usecases is creating a large model on meta device for further fx tracing:

1. create a large model on meta device on master node which doesn't have enough memory to materialize the model.
2. trace and split it to submodules using PiPPy
3. send each submodule(on meta device) over RPC to the corresponding workers
4. move those submodules to cuda device(materialize/initialize probably using torchdistx)




[ghstack-poisoned]

jamesr66a reviewed (approved):
This looks fine to me, I'm going to stamp this pending veto by someone more familiar with the PT RPC+tensorpipe integration. Otherwise, probably fine to merge

@cbalioglu We need a way to transmit type/shape information across the wire for the purposes of shape/type specialized tracing on the RPC target. Currently, the way that is available in PT core (i.e. with the fewest dependency additions) is `MetaTensor`. Maybe we can explore porting to FakeTensor after it has been merged into core and proven out.

mrshenli reviewed (approved):
Stamp to unblock

pbelevich committed:
Update on "RPC Meta device support"

We want to support meta device in PyTorch RPC framework. One of possible usecases is creating a large model on meta device for further fx tracing:

1. create a large model on meta device on master node which doesn't have enough memory to materialize the model.
2. trace and split it to submodules using PiPPy
3. send each submodule(on meta device) over RPC to the corresponding workers
4. move those submodules to cuda device(materialize/initialize probably using torchdistx)




[ghstack-poisoned]

pbelevich committed:
Update on "RPC Meta device support"

We want to support meta device in PyTorch RPC framework. One of possible usecases is creating a large model on meta device for further fx tracing:

1. create a large model on meta device on master node which doesn't have enough memory to materialize the model.
2. trace and split it to submodules using PiPPy
3. send each submodule(on meta device) over RPC to the corresponding workers
4. move those submodules to cuda device(materialize/initialize probably using torchdistx)




[ghstack-poisoned]

pbelevich committed:
Update on "RPC Meta device support"

We want to support meta device in PyTorch RPC framework. One of possible usecases is creating a large model on meta device for further fx tracing:

1. create a large model on meta device on master node which doesn't have enough memory to materialize the model.
2. trace and split it to submodules using PiPPy
3. send each submodule(on meta device) over RPC to the corresponding workers
4. move those submodules to cuda device(materialize/initialize probably using torchdistx)




[ghstack-poisoned]

pbelevich committed:
Update on "RPC Meta device support"

We want to support meta device in PyTorch RPC framework. One of possible usecases is creating a large model on meta device for further fx tracing:

1. create a large model on meta device on master node which doesn't have enough memory to materialize the model.
2. trace and split it to submodules using PiPPy
3. send each submodule(on meta device) over RPC to the corresponding workers
4. move those submodules to cuda device(materialize/initialize probably using torchdistx)




[ghstack-poisoned]

pbelevich commented:
@pytorchbot merge this please
